### PR TITLE
Replace deprecated and internal Skia EncodedImage calls with public versions

### PR DESCRIPTION
**Pull request by kjlubick**
Skia would like to remove `SkImageGenerator::MakeFromEncoded` and this appears to be the only remaining usage. It appears to be easily swapped out for the `SkImages::DeferredFromEncodedData`. (skbug.com/13052)

This also removes the use of the internal `SkCodecImageGenerator` for the public `SkCodec` API (which the image generator had just been deferring to anyway).

While unbreaking some unit tests, I made a few assertions easier to debug and produce nicer error messages.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by kjlubick** [comment]
Looks like Skia will need to expose https://github.com/google/skia/blob/main/src/codec/SkPixmapUtils.h#L37 or something similar to make some of the tests like `ImageDecoderTest.VerifySubpixelDecodingPreservesExifOrientation` continue to pass, as that was something the SkCodecImageGenerator handled

**Comment by kjlubick** [comment]
Hopefully I've found all the places where the SkCodecImageGenerator was transparently applying Exif metadata rotations and made them explicit on the Flutter-side.

File: impeller/image/backends/skia/compressed_image_skia.cc [hunk 1]
```diff
@@ -46,12 +47,12 @@ DecompressedImage CompressedImageSkia::Decode() const {
       },
       src);
 
-  auto generator = SkImageGenerator::MakeFromEncoded(sk_data);
-  if (!generator) {
+  auto img = SkImages::DeferredFromEncodedData(sk_data);
```
Comment (dnfield) [inline impeller/image/backends/skia/compressed_image_skia.cc:50]: nit: please spell out `image` here and below.

File: lib/ui/painting/image_generator.cc [hunk 2]
```diff
@@ -133,7 +148,39 @@ bool BuiltinSkiaCodecImageGenerator::GetPixels(
   if (prior_frame.has_value()) {
     options.fPriorFrame = prior_frame.value();
   }
-  return codec_generator_->getPixels(info, pixels, row_bytes, &options);
+  SkEncodedOrigin origin = codec_->getOrigin();
+
+  SkPixmap dst(info, pixels, row_bytes);
+  SkPixmap tmp;
+  SkBitmap tmpBitmap;
```
Comment (dnfield) [inline lib/ui/painting/image_generator.cc:155]: nit: naming - flutter uses snake_case for variable names, and prefers to spell things out (although dst/tmp are pretty commonly used).

**Review by dnfield — APPROVED** [review]
LGTM with a couple style nits.

File: lib/ui/painting/image_generator.cc (hunk 2 above)
Comment (kjlubick) [reply lib/ui/painting/image_generator.cc:155]: Done.

File: impeller/image/backends/skia/compressed_image_skia.cc (hunk 1 above)
Comment (kjlubick) [reply impeller/image/backends/skia/compressed_image_skia.cc:50]: Done

**Comment by chinmaygarde** [comment]
I believe this is good to go. Rebasing to kick off CI which tripped on a timeout.

**Review by chinmaygarde — APPROVED** [review]

**Comment by dnfield** [comment]
The google failures are not because of this changed, filed b/280066465 and manually marked status.

**Pull request merged**